Count issue with related models using though model #1795

regevbr opened the pull request:
Fixes #1795 - fixed wrong behaviour of `count`, `findOne`, `updateAll`, `destroyAll` on related models that use through models.

Fixes #1796 - Optimized the bloated query being used on through models when performing `find`, making it more efficient

Added optimization when querying related models using the id in the filter, to make the call to the through model more efficient.

Applied the fix from strongloop/loopback/issues/1076 to the explicit call of `find` on a related model


## Checklist
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](https://loopback.io/doc/en/contrib/style-guide-es6.html)
- [x] Commit messages are following our [guidelines](https://loopback.io/doc/en/contrib/git-commit-messages.html)

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

regevbr commented:
@bajtos can you please review?
I added 4 tests that failed before the fix (TDD style) and fixed all problematic functions - count,findOne,updateAll and destroyAll

regevbr commented:
@bajtos I combined here the fix of #1796 as well and they are closely related and  they use the same logic.

dhmlau commented:
@slnode test please

bajtos commented:
Thank you @regevbr for the pull request. This is a large change, I need to set aside some time to review it in detail. I'll try to get it done by the end of next week, before I leave for vacation.

@raymondfeng can you please take a look too?

bajtos commented on test/relations.test.js:
```diff
@@ -605,6 +605,864 @@ describe('relations', function() {
+    describe('scoped queries', function() {
+      describe('find', function() {
+        it('should allow to use fields on related model', function(done) {
+          Physician.create(function(err, physician) {
```
Unhandled `err`.

Please use async/await style in new tests, juggler supports only Node.js 8+ version now.

```ts
const physician = await Physician.create();
```

bajtos replied:
(The same comment applies to all other places where you are invoking a function or a method and providing a callback.)

regevbr replied:
Fixed

bajtos commented on test/relations.test.js:
```diff
@@ -605,6 +605,864 @@ describe('relations', function() {
+        it('should find scoped record with promises based on related model id', function(done) {
+          let id;
+          Physician.create()
+            .then(function(physician) {
```
Similarly here. Please use `await` instead of Promise/`.then()` style.

```ts
const physician = await Physician.create();
```

bajtos replied:
(The same comment applies to all other places where you are using `.then()` or `.catch()`.)

regevbr replied:
Fixed

regevbr commented:
@bajtos sure I will try to get on it soon, I have a lot on my plate :-)

regevbr commented:
@bajtos I fixed the tests (which actually got me to find bugs I introduced so I fixed them as well)

bajtos commented:
Thank you @regevbr for the updates, the tests look much better now!

I am afraid I got stuck in other tasks and did not find enough time to review your changes. If nobody from @strongloop/loopback-3x-maintainers and/or @strongloop/loopback-maintainers steps up to review this pull request, then I am afraid you will have to wait until January.

Sorry for the delay 😢

bajtos commented:
@slnode test please

bajtos commented:
Sorry @regevbr for the long silence on our side. The pull request is quite involved, I find it difficult to find enough time (and energy) to dive into the changes. Is there any chance to split this patch into smaller pull requests, where each pull request would be a meaningful change on its own? The smaller the patch, the easier it is to review for us.
